### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by jonasrlg**
1. Add README and Jupyter Notebooks;
2. Create Gradient Descent algorithm and functions for unnormalized PCs;
3. Change structure of `update` functions.

### CODE REVIEW

File: src/Learning/Gradient.jl [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  if learngaussians
+    Threads.@threads for i ∈ 1:length(gaussiannodes)
+      g = gaussiannodes[i]
+      G = curr[g]
+      # Gaussian scope
+      X = view(Data, :, G.scope)
+      # Gaussian parameters
+      μ, σ = G.mean, G.variance
+      # Δμ = 1/N ∑ (x-μ)/σ f(x)
+      Δμ = sum(((X .- μ) ./ σ) .* view(V, :, g))/numrows
```
Comment (RenatoGeh) [inline src/Learning/Gradient.jl:105]: `V` is in logspace, and if I understand what is happening here, `(X .- μ) ./ σ` should be part of the derivative and not in logspace. Am I missing something or misunderstanding the equation?

File: src/Learning/Gradient.jl [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      G.mean -= learningrate*Δμ
+      # Derivate (w.r.t σ) of log f(x)
+      log_d = ((X .- μ).^2)./ (2*σ^2) .- 1/(2*σ)
+      # Δσ = 1/N ∑ [(x-μ)^2/(2z^2) - 1/(2*σ)] f(x)
+      Δσ = sum(log_d .* view(V, :, g))/numrows
```
Comment (RenatoGeh) [inline src/Learning/Gradient.jl:110]: The same thing here, except here you signaled this as `log_d` and are multiplying with `V`. If `log_d` is the derivative in logspace shouldn't this be a `.+`? The sum (and `1/N`) should only occur in non-logspace, right? This also happens in the mean case.

File: src/Learning/Gradient.jl [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      G.variance -= learningrate*Δσ
+      !(G.variance > minimumvariance) && (G.variance = minimumvariance)
+    end
+  end
+
+  if verbose && (learner.steps % 100 == 0)
+    # TODO: Mudar o nome de LL para outra coisa que represente S(d), valor do circuito no dado
```
Comment (RenatoGeh) [inline src/Learning/Gradient.jl:117]: You may call this the objective (function) score which we're trying to maximize.

File: src/Learning/Gradient.jl [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    if m <= numrows
+      ll = mLL!(view(V, 1:m, :), curr, validation) - norm_const
+    else
+      ll = mLL!(Matrix{Float64}(undef, m, length(curr)), curr, validation) - norm_const
+    end
+    if !isnothing(history) push!(history, ll) end
+    println("Iteration $(learner.steps). η: $(learningrate), LL: $(ll)")
+  end
+
+  swap!(learner.circ)
```
Comment (RenatoGeh) [inline src/Learning/Gradient.jl:130]: You don't need this for the gradient. This was only needed for EM and mainly for batch learning.

File: src/Learning/Gradient.jl [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  smoothing::Float64 = 1e-4,
+  learngaussians::Bool = false,
+  minimumvariance::Float64 = learner.minimumvariance,
+  verbose::Bool = false, validation::AbstractMatrix = Data, history = nothing
+)
+
+  numrows, numcols = size(Data)
+
+  curr = learner.circ.C
+  prev = learner.circ.P
```
Comment (RenatoGeh) [inline src/Learning/Gradient.jl:57]: `prev` is not needed for your implementation.

File: src/Learning/Parameters.jl [hunk 6]
```diff
@@ -173,6 +173,51 @@ function backpropagate_tree!(Δ::AbstractMatrix{<:Real}, C::Vector{Node}, V::Abs
   return nothing
 end
 
+```
+  Marginalized version of backpropagate_tree.
+```
+function norm_backpropagate_tree!(marg_Δ::AbstractVector{<:Real}, C::Vector{Node}, marg_V::AbstractVector{<:Real})
```
Comment (RenatoGeh) [inline src/Learning/Parameters.jl:179]: This is paralellizable, I'm sure.

File: src/Learning/Parameters.jl [hunk 7]
```diff
@@ -173,6 +173,51 @@ function backpropagate_tree!(Δ::AbstractMatrix{<:Real}, C::Vector{Node}, V::Abs
...
+  marg_Δ .= 0.0
```
Comment (RenatoGeh) [inline src/Learning/Parameters.jl:180]: Not all of the matrix needs to be set to zero. See https://github.com/RenatoGeh/RPCircuits.jl/blob/a4cfef7c92c324aa4252f8c7ed834d1d6ac4e4af/src/Learning/Parameters.jl#L66 for more information. We only need to set to zero entries which are either incremented throughout the loop or belong to the root.

File: src/Learning/Parameters.jl [hunk 8]
```diff
@@ -173,6 +173,51 @@ function backpropagate_tree!(Δ::AbstractMatrix{<:Real}, C::Vector{Node}, V::Abs
...
+          marg_Δ[c] += N.weights[j] * exp(d)
+        else
+          marg_Δ[c] = log(N.weights[j]) + d
+        end
+      end
+    elseif isprod(N)
+      v = marg_V[i]
+      for c ∈ N.children
+        δ = d + v - marg_V[c]
+        isinvalid(δ)
```
Comment (RenatoGeh) [inline src/Learning/Parameters.jl:203]: I think your code is based on an older version of `backpropagate`. Check the newer, improved and (hopefully) bug free current version here https://github.com/RenatoGeh/RPCircuits.jl/blob/a4cfef7c92c324aa4252f8c7ed834d1d6ac4e4af/src/Learning/Parameters.jl#L104.

**Review by RenatoGeh — CHANGES_REQUESTED** [review]
Thanks for the PR, Jonas. There are some issues I would like you to address, but overall the PR is looking good. There are a few typos here and there, but I'll deal with them when I manually merge.
